src/main/java/org/zalando/nakadi/service/subscription/SubscriptionTimeLagService.java

```diff
@@ -0,0 +1,129 @@
+
+    private static final int EVENT_FETCH_WAIT_TIME_MS = 1000;
+    private static final int COMPLETE_TIMEOUT_MS = 30000;
+    private static final int LAG_CALCULATION_PARALLELISM = 10;
+
+    private final TimelineService timelineService;
+    private final NakadiCursorComparator cursorComparator;
+
+    @Autowired
+    public SubscriptionTimeLagService(final TimelineService timelineService,
+                                      final NakadiCursorComparator cursorComparator) {
+        this.timelineService = timelineService;
+        this.cursorComparator = cursorComparator;
+    }
+
+    public Map<EventTypePartition, Duration> getTimeLags(final Collection<NakadiCursor> committedPositions,
+                                                         final List<PartitionEndStatistics> endPositions)
+            throws ErrorGettingCursorTimeLagException, InconsistentStateException {
+
+        final ExecutorService executor = Executors.newFixedThreadPool(LAG_CALCULATION_PARALLELISM);
+
+        final Map<EventTypePartition, Future<Duration>> futures = new HashMap<>();
+        final Map<EventTypePartition, Duration> result = new HashMap<>();
+
+        for (final NakadiCursor cursor : committedPositions) {
+            final EventTypePartition partition = new EventTypePartition(cursor.getTimeline().getEventType(),
+                    cursor.getPartition());
+            if (isCursorAtTail(cursor, endPositions)) {
+                result.put(partition, Duration.ZERO);
+            } else {
+                futures.put(partition, executor.submit(() -> getNextEventTimeLag(cursor)));
+            }
+        }
+
+        executor.shutdown();
+        try {
+            final boolean finished = executor.awaitTermination(COMPLETE_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+            if (!finished) {
+                throw new InconsistentStateException("Timeout occurred when getting subscription time lag");
+            }
+            for (final EventTypePartition partition : futures.keySet()) {
+                final Future<Duration> future = futures.get(partition);
+                try {
+                    result.put(partition, future.get());
+                } catch (final ExecutionException e) {
+                    throw e.getCause();
+                }
+            }
+            return result;
+
+        } catch (final InconsistentStateException | ErrorGettingCursorTimeLagException e) {
+            throw e;
+        } catch (final Throwable e) {
+            throw new InconsistentStateException("Unexpected error occurred when getting subscription time lag", e);
+        }
+    }
+
+    private Duration getNextEventTimeLag(final NakadiCursor cursor) throws ErrorGettingCursorTimeLagException,
+            InconsistentStateException {
+        try {
+            final EventConsumer consumer = timelineService.createEventConsumer(
+                    "time-lag-checker-" + UUID.randomUUID().toString(), ImmutableList.of(cursor));
```

Comment:
I did not even remember this "identification" feature still existed. I thought it was "commented out" as an emergency reaction to an incident which we never followed up.

Reply:
It is commented out in a place where this parameter is passed to kafka consumer.